### PR TITLE
Load language strings for a default language to a separate array

### PR DESCRIPTION
**Pull request by mbabker**
### Summary of Changes

Right now within `Joomla\CMS\Language\Language` when loading a default language's strings, those strings are mixed in with the strings of the active language.  While the behavior is technically correct, this results in two message catalogues being mixed into one with data being overwritten during the various calls.  Instead of doing this, a more preferential behavior is to treat each language's strings as a separate message catalogue, and that is what this PR implements.

A new class property, `$defaultStrings`, is introduced which is similar to the existing `$strings` array except it will only hold the strings of the default language when those are loaded, resulting in the `$strings` array now only holding strings of the active language.  Methods which do read/write operations to this array (the `_` method handling converting language keys to translated strings, the `load` and `loadLanguage` methods when reading files into memory, and the `hasKey` method when checking for existence of a key) are updated to work with this new methodology of splitting strings into different properties.

### Testing Instructions

Install a second language to your Joomla site and make that your active language.  With this PR applied, the translation keys should continue to be displayed in your active language if available, fall back to English if not available (when debugging is not active, when active we will skip this check), and display the untranslated key if not present at all.

### Documentation Changes Required

If accepted, anyone who has subclassed this class and is working with the `$strings` class property should update their code to be aware of the new `$defaultStrings` class property.  If they don't, the worst thing that happens is they would no longer get the English language fallbacks as those won't be included in the existing property.

### CODE REVIEW

**Comment by mbabker** [comment]
@Bakual this is what I was referencing in the other issue.

**Comment by ggppdk** [comment]
A little more memory (compared to total memory usage) is worthwhile for this change

and since this will go to a major version change 3.7 -> 3.8  ?
if any 3rd party language-related extension need some updating, they will have time to do it

**Comment by mbabker** [comment]
> and since this will go to a major version change 3.7 -> 3.8 ?

1) We aren't WordPress, that's not a major version 😉 
2) Technically it is fully B/C, unless you really want to start getting into semantics about how much data is stored in one class property (which I don't believe is covered under any sane B/C definition)

**Comment by Bakual** [comment]
Looks nice from what I see and understand 👍

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 98539de0d272dc31637fc11dac06b82fdb815ca6<br /><br />Behavior after this patch is the same as before here and as detailed in the test instructions. No specific issue.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17398">issues.joomla.org/tracker/joomla-cms/17398</a>.</sub>

**Comment by brianteeman** [comment]
@mbabker can you fix the conflicts please

**Comment by ghost** [comment]
Reminder @mbabker for solving conflicting Files so this PR can be tested.

**Comment by mbabker** [comment]
Rebased onto 3.9 branch with conflicts resolved.

File: libraries/src/Language/Language.php [hunk 1]
```diff
@@ -347,6 +364,24 @@ public function _($string, $jsSafe = false, $interpretBackSlashes = true)
 				$this->used[$key][] = $caller;
 			}
 		}
+		// Only look in the default store if debugging is not active to allow a key to be reported as untranslated in the active language
+		elseif (!$this->debug && isset($this->defaultStrings[$key]))
+		{
+			$string = $this->debug ? '**' . $this->defaultStrings[$key] . '**' : $this->defaultStrings[$key];
```
Comment (ReLater) [inline libraries/src/Language/Language.php:370]: When can `$this->debug` be true here after the `elseif (!$this->debug && ...` condition in line 368?

File: libraries/src/Language/Language.php [hunk 2]
```diff
@@ -347,6 +364,24 @@ public function _($string, $jsSafe = false, $interpretBackSlashes = true)
...
+
+			// Store debug information
+			if ($this->debug)
```
Comment (ReLater) [inline libraries/src/Language/Language.php:373]: When can `$this->debug` be true here after the `elseif (!$this->debug && ...` condition in line 368?

**Comment by mbabker** [comment]
I'm abandoning this.  Working on some stuff in the Framework package (having the message catalogue extracted to a separate object with the catalogue having its own awareness of configurable fallback catalogues) that this approach conflicts with.

**Pull request closed without merging**